packages/marshal/src/types.js

```diff
@@ -2,68 +2,334 @@
+ * nominal types in a structural type system.
  *
- * A Passable has a pass-by-copy superstructure. This includes the atomic
- * pass-by-copy primitives ("bigint" | "boolean" | "null" | "number" |
- * "string" | "undefined") and the composite pass-by-copy objects ("copyArray" |
- * "copyRecord" | "copyError"). The composite pass-by-copy objects that may
- * contain other Passables.
+ * TODO We can use this pseudo-property to obtain the `PrimaryRemotable` type.
+ * We can then extend the definition of `E()` to deal with `Remote<T>`s as if
+ * they have the methods and properties found on `T`.
  *
- * A Passable's pass-by-copy superstructure ends in PassableCap leaves. The
- * Passable can be further classified by the nature of these leaves. Since a
- * Passable is hardened, its structure and classification is stable --- its
- * structure and classification cannot change even if some of the objects are
- * proxies.
+ * @template {PrimaryRemotable} T
+ * @typedef {{_Remote: T}} Remote
+ * A `Remote<T>` is a Remotable object in one vat that represents a
+ * PrimaryRemotable of type T in another vat. The Remote has an unforgeable
+ * identity that is one-to-one with the unforgeable identity of its primary.
+ * If there is only one comm system between the two vats, then the identity of
+ * the Remote is not observably different than the identity of its primary.
+ *
+ * A Remote, being a kind of Remotable, is pass-by-remote. When a Remote<T>
+ * is passed as an argument in a message to an object in another vat, it
+ * arrives as either a T or a Remote<T>, depending on whether the destination
+ * is the home vat of its primary.
  */
 
 /**
- * @typedef {Passable} Comparable
+ * @template {PrimaryRemotable} T
+ * @typedef {T | Remote<T>} Remotable
+ * Corresponding PassStyle is "remotable"
+ *
+ * A Remotable<T> is either the PrimaryRemotable T in its home vat, or is one
+ * of its Remote<T> representatives in other vats. The primary together with
+ * all of its remotes are, in aggregate, the abstact distributed Remotable<T>
+ * object consisting of all these individual remotables. If there is only one
+ * coherent comm system among the vats in question, then the remotable as a
+ * whole effectively has the unforgeable identity of its primary, since the
+ * identity of each of its remotables is not observably different than the
+ * identity of its primary. (TODO explain relationship to the Unum model.)
+ *
+ * A Remotable is a PassableKey, which `sameKey` compares for equivalence
+ * according to its unforgeable identity, which effectively compares the
+ * unforgeable identities of their primaries.
  *
- * A Comparable is a Passable in which none of the leaves of the pass-by-copy
- * superstructure are promises. Two Comparables may be compared by
- * for equivalence according to `sameStructure`, which is the strongest
- * equivalence class supported by marshal's distributed object semantics.
+ * Remotables are unordered and cannot be sorted.
```

Comment:
This makes sense. Just adding a note here that the implication is that for an `amount` with a `SetValue` that is an array of opaque empty Remotables (handles) (or potentially a CopySet of handles), we have no way of cutting down a comparison of two arrays of size m and n from O(m*n). To be clear, that's the current scenario.

Reply:
Ah, wait, we might be able to do better with CopySets, depending on what the operation is.